### PR TITLE
Support user subclass before save

### LINKED ISSUE #25: Get fatuhoku pr's merged...
@fatuhoku  has three outstanding pr's from December.  The only thing outstanding is to squash, which hasn't happened in six months and is preventing the mainline from being as useful as it could be.  I think that there are a few options:

1. use the github squash on merge options.  I haven't used, but i think it'll just do the right thing here
2. i could do the squash and try to maintin @fatuhoku as the author, and then open up three new prs
3. close the prs to indicate that they are not going to get merged
4. merge them as is. i for one would think the value to the package would outweigh the downside of adding more commits to the history....
 
I am going to mege his pr's into my master and change my npm to pull my master.  if i then make any changes (like adding a mock response to the before save) they wont be able to be merged in to this repo
 :(.

### PR DESCRIPTION
**Pull request by acinader**
fixes #25 

unit test coverage for _User class

PS the third commit 75c99d2152733ac5cd3fccc5e6ab4996cc75b2b4 just adds semicolons.  was bothering me and i figured out linting....

### CODE REVIEW

File: src/parse-mockdb.js [hunk 1]
```diff
@@ -112,8 +112,7 @@ function runHook(className, hookType, data) {
       object: model,
       user: "ParseMockDB doesn't define request.user."
     };
-
-    return hook(beforeSaveOrBeforeDeleteRequestObject).bind(model)().then((beforeSaveOverrideValue) => {
+    return hook(beforeSaveOrBeforeDeleteRequestObject).then((beforeSaveOverrideValue) => {
```
Comment (acinader) [inline src/parse-mockdb.js]: i wish i understood what the .bind(model)() was supposed to do and if it should be there.  but now the unit tests pass.

:(  any insight you guys can give me as to what it was supposed to do, ideally in the form of a failing test so i can try to figure it out.....

**Comment by acinader** [comment]
@TylerBrock this is ready for review.

**Comment by acinader** [comment]
@fatuhoku take a look.  I rebased your changes, and i took a look at what you did in handle_request, but when i factored it out, i found that it is exactly the same as what was there before?

I fixed the unit tests, but I now need a unit test that fails, cause this really looks like a noop to me?  maybe something was fixed in between and when i rebased it, everything just worked??

**Comment by acinader** [comment]
@TylerBrock I think that this can be closed unless you think that the extra unit tests are worth keeping?  If so, I'll rebase.

**Comment by acinader** [comment]
@fatuhoku I hope you'll try head again and if there is any functionality that you've got implemented that isn't in head, I'd be glad to work with you to get it worked into head!!!

**Comment by TylerBrock** [comment]
@fatuhoku agree completely with @acinader thank you for all the hard work. Please continue to contribute.

@acinader what is your opinion on keeping the tests? worth it?

**Comment by acinader** [comment]
@TylerBrock probably?  

```
npm test | grep ✓ | awk '{print substr($0, index($0, $3))}' | sort | uniq -c

2 rejects the delete if there is a problem
2 rejects the save if there is a problem
2 runs the hook before deleting the object
2 runs the hook before saving the model and persists the object
....
```
it's certainly valid that those tests should be run twice, once for a "plain" parse class, and once for the special parse class _User, but the code shouldn't be duplicated.  I'll see if can quickly refactor, if so I'll amend this.  if not, we should probably merge it anyway :).

**Comment by acinader** [comment]
combined the tests for parse objects and _User using @fatuhoku methods.  Nice!

yup, @TylerBrock I'd like you to review it for merge.

**Comment by acinader** [comment]
@TylerBrock nudge

**Comment by TylerBrock** [comment]
Hey sorry for the delay. I'm out of town this weekend. Most likely I'll review on Tuesday.

**Comment by TylerBrock** [comment]
I don't think this is mergeable if the deep save test fails. Can we figure out what is going on there?

**Comment by acinader** [comment]
i didn't touch that test https://github.com/HustleInc/parse-mockdb/blob/master/test/test.js#L830

it fails on master too.  the only thing i did was change the test from being commented out to marked skip using ``xit``.

which is nice because it shows that the test is skipped when you npm test:

```  
    ✓ should skip and limit items appropriately
    - should deep save and update nested objects
    ✓ successfully uses containsAll query
```

The test's comment cites an issue that you had opened: https://github.com/ParsePlatform/Parse-SDK-JS/issues/91

the issue was marked resolved in parse sdk in version 1.6.14.  I'm running 1.8.5 as our package.json has  ``>=1.6.0``

I can open an issue on this repo and look into why it still fails, any background you have might be helpful?

**Comment by TylerBrock** [comment]
Oh, duh, thank you! Much better! It was too early to be looking at diffs apparently.

Yeah, it would be great to get an issue to track it. Would be happy to fill in info where necessary.

**Comment by TylerBrock** [comment]
Two questions:
 - want me to do a release for this change?
 - should we update the readme to reflect the current state of the project?

**Comment by acinader** [comment]
Yeah the readme and some code comments need updating.  Will do.

**Comment by acinader** [comment]
no opinion on release. 

here's an update for the readme: #41

**Pull request merged**